### PR TITLE
Remove dependency on k8s.io/kubernetes

### LINKED ISSUE #5041: Remove dependency on k8s.io/kubernetes
**Is your feature request related to a problem? Please describe.**

I'm working on a feature in external-dns and gloo's dependency on `k8s.io/kubernetes` is preventing me from being able to import gloo's API.

**Describe the solution you'd like**

The dependency headaches from importing `k8s.io/kubernetes` are vast and gloo's usage of it is small. It can easily be removed.

**Describe alternatives you've considered**

The current solution in external-dns is to use a dynamic Kubernetes client, re-implement the gloo API internally, and marshal the necessary objects from unstructured data. It's brittle and causing problems with newer versions of the Kubernetes test/mock client.

### PR DESCRIPTION
**Pull request by abursavich**
# Description

This fixes #5041 and removes the dependency on `k8s.io/kubernetes`.

# Context

The dependency on `k8s.io/kubernetes` is preventing me from being able to import gloo's API, because of the pain that comes along with it.

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5041

### CODE REVIEW

**Comment by kdorosh** [comment]
thanks for the contribution! /test

**Comment by abursavich** [comment]
From the error message, it think the build-bot is using a pre-1.16 version of Go. Can we bump that up? The other test may fail similarly, depending on their Go versions.

**Comment by kdorosh** [comment]
we should be able to bump that. if you update the code (https://github.com/solo-io/build-bot/blob/master/go.mod#L3) i can deploy the new bot in our environment

**Comment by sam-heilbron** [comment]
Is it a build-bot issue, or because solo-apis requires client-go which is now excluded?
https://github.com/solo-io/solo-apis/blob/master/go.mod#L37

**Comment by abursavich** [comment]
> Is it a build-bot issue, or because solo-apis requires client-go which is now excluded?

It's both. The way excludes work changed in Go 1.16. With Go 1.15, those excludes break builds in the way you're suggesting. With Go 1.16, those excludes make the build work without needing to require the wrong version and use replace hacks to fix it.

**Comment by saiskee** [comment]
Hi @abursavich, could you link us to the go 1.16 docs for this change? This behavior checks out from local testing, but it would be good to have some official docs, which I'm unable to find currently.

**Comment by abursavich** [comment]
From the [1.6 release notes](https://golang.org/doc/go1.16):

> The go command now ignores requirements on module versions excluded by exclude directives in the main module. Previously, the go command used the next version higher than an excluded version, but that version could change over time, resulting in non-reproducible builds.

**Comment by sam-heilbron** [comment]
Hi @abursavich, I'm going to do some reading to better understand the scope of this change. This type of change to build-bot could/would affect others teams within the organization.

**Comment by abursavich** [comment]
No worries. I'm not in any rush.

**Comment by abursavich** [comment]
It's been about 2 months. Please let me know if anyone's gonna take a look at this. I'll probably wait until then to resolve any conflicts.

**Comment by abursavich** [comment]
Alright... I rebased it once.

**Comment by sam-heilbron** [comment]
Apologies for the delay. I'm doing some tangential work that I hope to resolve this. It's taken a little longer than expected. I will update this issue when the work is completed

**Comment by sam-heilbron** [comment]
build-bot failing with: `go: github.com/solo-io/solo-apis@v0.0.0-20210922150112-505473b2e66c: github.com/solo-io/solo-apis(v0.0.0-20210922150112-505473b2e66c) depends on excluded k8s.io/client-go(v11.0.0+incompatible) with no newer version available`

**Comment by abursavich** [comment]
@sam-heilbron, is the build bot still using a pre-1.16 version of go?

We did this same thing ~2 months ago: https://github.com/solo-io/gloo/pull/5042#issuecomment-882661911

**Comment by abursavich** [comment]
I'm not sure if I'm reading the logs correctly, but it _looks_ like it's using v1.13.5.

```
$ docker run --entrypoint go gcr.io/solo-public/prepare-go-workspace:0.3.0 version
go version go1.13.5 linux/amd64
```

**Comment by sam-heilbron** [comment]
/kick https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=887a7d82-54b1-42e2-8722-eb55720f13d7

**Review by sam-heilbron — CHANGES_REQUESTED** [review]
Discussed offline that this will go into our 1.10.0-beta1 release, since we're too close to the 1.9 release. Requesting changes to ensure it doesn't get merged accidentally before we cut a new branch

File: go.mod [hunk 1]
```diff
@@ -139,36 +136,35 @@ replace (
...
-	k8s.io/client-go => k8s.io/client-go v0.19.6
-	k8s.io/cloud-provider => k8s.io/cloud-provider v0.19.6
-	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.19.6
-	k8s.io/code-generator => k8s.io/code-generator v0.19.6
-	k8s.io/component-base => k8s.io/component-base v0.19.6
-	k8s.io/cri-api => k8s.io/cri-api v0.19.6
-	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.19.6
-	k8s.io/gengo => k8s.io/gengo v0.0.0-20190822140433-26a664648505
-	k8s.io/heapster => k8s.io/heapster v1.2.0-beta.1
+	google.golang.org/grpc => google.golang.org/grpc v1.36.0
```
Comment (sam-heilbron) [inline go.mod:139]: this upgrade is being handled in a separate PR: https://github.com/solo-io/gloo/pull/5429

**Review by MitchAman — COMMENTED** [review]
LGTM

**Review by sam-heilbron — APPROVED** [review]

**Pull request merged**